scripts/prune-dts/prune-dts.ts

```diff
@@ -0,0 +1,436 @@
+ * class hierarchies for classes whose inherited types are not part of the
+ * public API.
+ *
+ * This method relies on a private API in TypeScript's `codefix` package.
+ */
+function convertPropertiesForEnclosingClass(
+  program: ts.Program,
+  host: ts.CompilerHost,
+  sourceFile: ts.SourceFile,
+  parentClassSymbols: ts.Symbol[],
+  currentClass: ts.ClassDeclaration | ts.InterfaceDeclaration
+): ts.NamedDeclaration[] {
+  const newMembers: ts.NamedDeclaration[] = [];
+  // The `codefix` package is not public but it does exactly what we want. We
+  // can explore adding a slimmed down version of this package to our repository
+  // if this dependency should ever break.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (ts as any).codefix.createMissingMemberNodes(
+    currentClass,
+    parentClassSymbols,
+    sourceFile,
+    { program, host },
+    /* userPreferences= */ {},
+    /* importAdder= */ undefined,
+    (c: ts.ClassElement) => newMembers.push(c)
+  );
+  return newMembers;
+}
+/**
+ * Replaces input types of public APIs that consume non-exported types, which
+ * allows us to exclude private types from the pruned definitions. Returns the
+ * the name of the exported API or undefined if no type is found.
+ *
+ * @example
+ * Input:
+ * class PrivateFoo {}
+ * export class PublicFoo extends PrivateFoo {}
+ * export function doFoo(foo: PrivateFoo);
+ *
+ * Output:
+ * export class PublicFoo {}
+ * export function doFoo(foo: PublicFoo);
+ */
+function extractExportedSymbol(
+  typeChecker: ts.TypeChecker,
+  sourceFile: ts.SourceFile,
+  typeName: ts.Node
+): ts.Symbol | undefined {
+  if (!ts.isIdentifier(typeName)) {
+    return undefined;
+  }
+
+  const localSymbolName = typeName.escapedText;
+  const allExportedSymbols = typeChecker.getExportsOfModule(
+    typeChecker.getSymbolAtLocation(sourceFile)!
+  );
+
+  // Examine all exported types and check if they extend or implement the
+  // provided local type. If so, we can use the exported type in lieu of the
+  // private type.
+
+  // Short circuit if the local types is already part of the public types.
+  for (const symbol of allExportedSymbols) {
+    if (symbol.name === localSymbolName) {
+      return symbol;
+    }
+  }
+
+  // See if there is an exported symbol that extends this private symbol.
+  // In this case, we can safely use the public symbol instead.
+  for (const symbol of allExportedSymbols) {
+    for (const declaration of symbol.declarations) {
+      if (
+        ts.isClassDeclaration(declaration) ||
+        ts.isInterfaceDeclaration(declaration)
+      ) {
+        for (const heritageClause of declaration.heritageClauses || []) {
+          for (const type of heritageClause.types || []) {
+            if (ts.isIdentifier(type.expression)) {
+              const subclassName = type.expression.escapedText;
+              if (subclassName === localSymbolName) {
+                return symbol;
```

Comment:
Just want to note that this is not always safe to replace a class with an inheriting class. Copied my comment from the previous PR:
> In case of multiple public classes extending `PrivateFoo`, we use the first one which seems arbitrary and is probably wrong - considering `PublicFoo1` and `PublicFoo2`, both of which extends `PrivateFoo`. `doFoo` should accept either of them, but after this change, it will only accept `PublicFoo1`.

Our API typings probably don't have the case I described, so the code works, but we should be aware that this transformation can't be generalized to work for any d.ts files.

Reply:
I think we probably need to return a Union type in this case. I added a TODO, and I can fix this in a follow-up PR. I do think that this scripts needs a few iterations before it works for all d.ts file.